### PR TITLE
Virtualbox Virtual Disk Image

### PR DESCRIPTION
**Pull request by KOLANICH**

### CODE REVIEW

**Comment by GreyCat** [comment]
Looks really cool! A few minor suggestions...

File: filesystem/vdi.ksy [hunk 1]
```diff
@@ -0,0 +1,228 @@
+meta:
+  id: vdi
+  endian: le
+  encoding: utf-8
+  application:
+    - VirtualBox
+    - QEMU
+    - VMWare Workstation
+  license: GPL3.0+
```
Comment (GreyCat) [inline filesystem/vdi.ksy]: 1. Probably something like `title: VDI (native VirtualBox Disk Image)` would be helpful.
2. Suggested order of `meta` entries is somewhat different in [style guide](http://doc.kaitai.io/ksy_style_guide.html).

File: filesystem/vdi.ksy [hunk 2]
```diff
@@ -0,0 +1,228 @@
...
+doc-ref: https://github.com/qemu/qemu/blob/master/block/vdi.c
+#  - https://github.com/qemu/qemu/blob/master/block/vdi.c
+#  - https://www.virtualbox.org/browser/vbox/trunk/src/VBox/Storage/VDI.cpp
+#  - https://forums.virtualbox.org/viewtopic.php?t=8046
```
Comment (GreyCat) [inline filesystem/vdi.ksy]: I'd rather put forums.virtualbox.org link here as a main one, as it is the essential source. Everything else (especially non-VirtualBox code) is reinterpretation — I'd mention them only if they contain any non-trivial clarifications / better info.

File: filesystem/vdi.ksy [hunk 3]
```diff
@@ -0,0 +1,228 @@
...
+    instances:
+      subheader_size_is_dynamic:
+        value: version.major>=1
+      header_size:
+        value: (subheader_size_is_dynamic?header_size_optional:336)
+      
+      bmap_size:
+        value: ((header_main.blocks_in_image * 4+header_main.geometry.sector_size-1) / header_main.geometry.sector_size)*header_main.geometry.sector_size
+      bmap_offset:
+        value: header_main.offset_bmap
```
Comment (GreyCat) [inline filesystem/vdi.ksy]: Probably it's better to use `ofs_` and `len_` prefixes consitently for stuff like that, as suggested in the style guide. Also, if you've decided to rename "bmap" into "blocks_map" (which I totally support), it should be consistent everywhere as well.

File: filesystem/vdi.ksy [hunk 4]
```diff
@@ -0,0 +1,228 @@
...
+
+      blocks_offset:
+        -orig-id: data_offset
+        value: header_main.offset_data
```
Comment (GreyCat) [inline filesystem/vdi.ksy]: Again, it's really harsh. If it's "blocks", then it should be "blocks" everywhere.

File: filesystem/vdi.ksy [hunk 5]
```diff
@@ -0,0 +1,228 @@
...
+        if: subheader_size_is_dynamic
+        
+      - id: header_main
+        size: header_size
+        type: header_main
+    instances:
+      subheader_size_is_dynamic:
+        value: version.major>=1
+      header_size:
+        value: (subheader_size_is_dynamic?header_size_optional:336)
```
Comment (GreyCat) [inline filesystem/vdi.ksy]: Spaces are welcome between operators and operands.

File: filesystem/vdi.ksy [hunk 6]
```diff
@@ -0,0 +1,228 @@
...
+              - id: heads
+                type: u4
+
+              - id: sectors
+                type: u4
+
+              - id: sector_size
+                type: u4
+          flags:
+            seq: #little endian assummed
```
Comment (GreyCat) [inline filesystem/vdi.ksy]: I believe VirtualBox is strictly little-endian anyway?

File: filesystem/vdi.ksy [hunk 7]
```diff
@@ -0,0 +1,228 @@
...
+              - id: fixed
+                type: b1
+              - id: reserved2
+                type: b8
+  blocks_map:
+    seq:
+      - id: index
+        type: block_index
+        repeat: expr
+        repeat-expr: _root.header.header_main.blocks_in_image
```
Comment (GreyCat) [inline filesystem/vdi.ksy]: `num_blocks` or something like that.

**Comment by GreyCat** [comment]
Thanks for this contribution!

File: filesystem/vdi.ksy (hunk 2 above)
Comment (KOLANICH) [reply filesystem/vdi.ksy]: I used QEMU source code as the main one: it is a bit simpler than vbox one. The names and docs are taken from there.  VBox source was used to look up some minor details like enum members and an additional geometry field.

File: filesystem/vdi.ksy (hunk 6 above)
Comment (KOLANICH) [reply filesystem/vdi.ksy]: The format is little endian, in qemu comments it is written that some fields are little endian even in memory.

File: filesystem/vdi.ksy (hunk 1 above)
Comment (KOLANICH) [reply filesystem/vdi.ksy]: It is speculated in multiple sources that VDI is an abbreviation for VirtualBox Disk Image.

**Pull request merged**